daffodil-lib/src/main/scala/org/apache/daffodil/util/CharacterSetRemapper.scala

```diff
@@ -0,0 +1,131 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.daffodil.util
+
+/**
+ * A abstract base for Remappers which convert strings.
+ *
+ * The public interface is just `def remap(s: String): String`.
+ *
+ * There are protected methods that implementations must provide.
+ *
+ * Contains shared implementation methods also.
+ *
+ * NOTE: This is inner loop stuff. Keep it and derived classes lean and fast.
+ * Use a java-like coding style. While loops, not map/flatmap/etc. avoid tuples.
+ */
+trait CharacterSetRemapper {
+
+  /**
+   * Remaps characters. Provides the previous and following characters since some remappings
+   * require this context for Surrogate Pairs, CRLF->LF etc.
+   *
+   * Plays a trick with negating the return value in order to avoid having to
+   * return more than one value, which is potentially less efficient.
+   *
+   * @param prev The character prior to the one being considered. (Needed for surrogates)
+   * @param curr The character under consideration for remapping.
+   * @param next The next character afterwards. (Needed for surrogates and CRLF pairs)
+   * @return The remapped character (as an Int) or that same remapped character Int
+   *         value negated, which signals that curr+next was remapped to a single character.
+   *         Such as is needed if CRLF is remapped to just LF.
+   */
+  protected def remap (prev: Int, curr: Int, next: Int): Int
+
+  /**
+   * Remaps the string. Returns the original string object if no remapping is required.
+   *
+   * Because of surrogate pairs, and the difference between 16-bit string codepoints
+   * and real character codes, lots of things that traverse strings need
+   * to consider either the codepoint after (if current is a leading surrogate)
+   * or codepoint before (if current is a trailing surrogate).
+   *
+   * This is not the only kind of character set remapping. In particular this is
+   * restricted to replace 1 character with 0 or 1 character in the remapped string.
+   * Other character set remappers can convert 1 character into 0 to many characters
+   * or even change from characters to bytes. Simple 1 to 1 remappings can be done
+   * with just a map, and 1 to N remappings can be done with flatmap if no context is
+   * needed for surrogates or CRLFs.
+   *
+   * See XMLUtils.walkUnicodeCharset for a more general kind of remapping that can
+   * replace 1 character with N as well as being context sensitive about adjacent
+   * characters before and after.
+   *
+   * This algorithm uses a StringBuilder which is not synchronized
+   * so it is noticably faster than StringBuffer, and since the StringBuilder
+   * is local to the function, we don't have to worry about any threading issues.
+   * This makes for a noticeable speed increase.
```

Comment:
Add to comment about inner-loop hence, this very Java coding style as performance matters for this.